### DISCUSSION TITLE
Picker values localization

### DISCUSSION THREAD

**Discussion by AncientLust** [dotnet/maui#25329] [Q&A] [answered]
What is a right way to localize picker values in MAUI? So far I have only one idea:

Use [LocalizationResourceManager.Maui](https://github.com/SirJohnK/LocalizationResourceManager.Maui). But for picker values it doesn't work when app localization has been changed until app restart.

```xaml
<Picker
    x:Name="NativeLanguagePicker"
    ItemsSource="{Binding NativeLanguageSetting.Options}"
    SelectedItem ="{Binding NativeLanguageSetting.Picked, Mode=TwoWay}"
    ItemDisplayBinding="{localization:TranslateBinding Localization, TranslateValue=True}"
    Style="{DynamicResource SettingPicker}">
</Picker>
```

Is there any way to do it right in MAUI? Because the whole app follows culture change and translates accordingly except of pickers. It looks like there should be a way to do it elegantly.

**Comment by stephenquan** [comment]
Interesting problem. I worked on TranslateBinding https://github.com/SirJohnK/LocalizationResourceManager.Maui/pull/16 but if there's an issue with it, I can definite look at fixing it. But, outside of TranslateBinding, the main issue is will the Picker respond to MVVM changes on the ItemDisplayBinding whilst the Picker is active? Having an answer to that question will help narrow down the problem. If you're not sure what I'm talking about. I could find time to create a repro case myself, but, it would be helpful if you are able to do that investigative work.

**Reply by AncientLust** [reply]
Unfortunately I don't have the exact answer, but I made a [repro project](https://github.com/AncientLust/PickerLocalizationIssue). I hope it will save you some time. Also attaching the video:

https://github.com/user-attachments/assets/4ef18b23-f7f8-498f-afb6-1c821bd39ef4

**Comment by stephenquan** [comment]
I believe the real issue is:

> Picker caches the values derived from ItemDisplayBinding

To work around the problem, I moved the TranslateBinding to your view model, by changing the Picker.ItemDisplayBinding to

```xaml
    ItemDisplayBinding="{Binding DisplayBinding}"
```

and then implementing DisplayBinding in your view model:

```c#
    public BindingBase DisplayBinding { get; } = new Binding("Localization", converter: new TranslateConverter());
    internal class TranslateConverter : IValueConverter
    {
        public object? Convert(object? value, Type targetType, object? parameter, CultureInfo culture)
            => (value is string s) ? PickerLocalizationIssue.Resources.Localization.AppResources.ResourceManager.GetString(s) : null;
        public object? ConvertBack(object? value, Type targetType, object? parameter, CultureInfo culture)
            => throw new NotImplementedException();
    }
```

Whenever you change languages I call `OnPropertyChanged(nameof(DisplayBinding));` to cause the Picker to update. I've got these changes in a PR: https://github.com/AncientLust/PickerLocalizationIssue/pull/1

**Comment by AncientLust** [comment]
Hi @stephenquan, just looked into your PR. Thank you for your time and effort! But when I run your PR on iOS device/emulator I see that pickers cannot pass initialization:

![IMAGE 2024-10-21 11:43:57](https://github.com/user-attachments/assets/aedf8d9b-1be9-4d3f-a37e-9c10f4f204c1)

If a go back to initial commit, I don't have this issue. Did you have successful test on your side?

I wouldn't really want to waste your time to fix it. Since I don't have enough competency to make myself the workaround you've shared, I hoped miss something simple. Because the whole situation looks a bit strange: the whole app can be localized with a single line of code except of pickers. But I gues we have what we have, it's juts not implemented on the design level.

**Comment by stephenquan** [comment, marked as answer]
Hi @AncientLust, I tested mainly on Windows. I will run an iOS test when I get a chance. In the meantime, I reverted my code to one of my earlier versions which, hopefully, would be more easier to edit.

Also, to improve localization support and to simplify the code-behind we have to write, I have raised two enhancements and their corresponding PR on the [LocalizationResourceManager.Maui](https://github.com/SirJohnK/LocalizationResourceManager.Maui) repo:
 - https://github.com/SirJohnK/LocalizationResourceManager.Maui/issues/34
 - https://github.com/SirJohnK/LocalizationResourceManager.Maui/issues/35
 - https://github.com/SirJohnK/LocalizationResourceManager.Maui/pull/33
